File: 프로그래머스/2/87946. 피로도/피로도.java
public class Solution {
    int answer = -1;
    public int solution(int k,int[][] dungeons) {
        
        int n = dungeons.length;
        boolean[] visited =  new boolean[n];
        backtrack(visited,n,k,dungeons,0);
        return answer;
    }

    public void backtrack(boolean[] visited, int n,int k, int[][] dungeons, int cnt) {
        if(cnt > answer) {
            answer = cnt;
        }
        //recursive call
        for(int i=0; i<n; i++) {
            if(k >= dungeons[i][0] && !visited[i]) {
                visited[i] = true;
                backtrack(visited,n,k-dungeons[i][1],dungeons,cnt+1);
                visited[i]=false;
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations
def solution(k, dungeons):
    answer = 0
    
    for order in permutations(dungeons) :
        remain = k
        count = 0
        
        for required, cost in order :
            if remain >= required :
                remain -= cost
                count += 1
            else :
                break
        answer = max(answer, count)
        
    return answer
    
    # 처음에 dungeons[0] >= k 이어야 시작 가능 -> 정렬을 해야하나???
    # 그 다음으로 올 수 있는 모든 경우의 수를 반복해야함...